src/Compilers/Core/Portable/SourceGeneration/SyntaxStore.cs

```diff
@@ -0,0 +1,130 @@
+        private readonly Compilation? _compilation;
+        internal static readonly SyntaxStore Empty = new SyntaxStore(StateTableStore.Empty, compilation: null);
+
+        private SyntaxStore(StateTableStore tables, Compilation? compilation)
+        {
+            _tables = tables;
+            _compilation = compilation;
+        }
+
+        public Builder ToBuilder(Compilation compilation, ImmutableArray<SyntaxInputNode> syntaxInputNodes, bool enableTracking, CancellationToken cancellationToken) => new Builder(compilation, syntaxInputNodes, enableTracking, this, cancellationToken);
+
+        public sealed class Builder
+        {
+            private readonly ImmutableDictionary<SyntaxInputNode, Exception>.Builder _syntaxExceptions = ImmutableDictionary.CreateBuilder<SyntaxInputNode, Exception>();
+            private readonly StateTableStore.Builder _tableBuilder = new StateTableStore.Builder();
+            private readonly Compilation _compilation;
+            private readonly ImmutableArray<SyntaxInputNode> _syntaxInputNodes;
+            private readonly bool _enableTracking;
+            private readonly SyntaxStore _previous;
+            private readonly CancellationToken _cancellationToken;
+
+            internal Builder(Compilation compilation, ImmutableArray<SyntaxInputNode> syntaxInputNodes, bool enableTracking, SyntaxStore previousStore, CancellationToken cancellationToken)
+            {
+                _compilation = compilation;
+                _syntaxInputNodes = syntaxInputNodes;
+                _enableTracking = enableTracking;
+                _previous = previousStore;
+                _cancellationToken = cancellationToken;
+            }
+
+            public IStateTable GetSyntaxInputTable(SyntaxInputNode syntaxInputNode, NodeStateTable<SyntaxTree> syntaxTreeTable)
+            {
+                Debug.Assert(_syntaxInputNodes.Contains(syntaxInputNode));
+
+                // when we don't have a value for this node, we update all the syntax inputs at once
+                if (!_tableBuilder.Contains(syntaxInputNode))
+                {
+                    // CONSIDER: when the compilation is the same as previous, the syntax trees must also be the same.
+                    // if we have a previous state table for a node, we can just short circuit knowing that it is up to date
+                    // This step isn't part of the tree, so we can skip recording.
+                    var compilationIsCached = _compilation == _previous._compilation;
+
+                    // get a builder for each input node
+                    var syntaxInputBuilders = ArrayBuilder<(SyntaxInputNode node, ISyntaxInputBuilder builder)>.GetInstance(_syntaxInputNodes.Length);
+                    foreach (var node in _syntaxInputNodes)
+                    {
+                        // TODO: We don't cache the tracked incremental steps in a manner that we can easily rehydrate between runs,
```

Comment:
Please replace TODO with a tracking issue

Reply:
Most of this is just a copy/paste move from existing code, but makes sense to clean it up as we do so.